vault/plugin_catalog.go

```diff
@@ -0,0 +1,169 @@
+package vault
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/vault/helper/builtinplugins"
+	"github.com/hashicorp/vault/helper/jsonutil"
+	"github.com/hashicorp/vault/helper/pluginutil"
+	"github.com/hashicorp/vault/logical"
+)
+
+var (
+	pluginCatalogPrefix = "plugin-catalog/"
+)
+
+// PluginCatalog keeps a record of plugins known to vault. External plugins need
+// to be registered to the catalog before they can be used in backends. Builtin
+// plugins are automatically detected and included in the catalog.
+type PluginCatalog struct {
+	catalogView *BarrierView
+	directory   string
+
+	lock sync.RWMutex
+}
+
+func (c *Core) setupPluginCatalog() error {
+	c.pluginCatalog = &PluginCatalog{
+		catalogView: c.systemBarrierView.SubView(pluginCatalogPrefix),
+		directory:   c.pluginDirectory,
+	}
+
+	return nil
+}
+
+// Get retrieves a plugin with the specified name from the catalog. It first
+// looks for external plugins with this name and then looks for builtin plugins.
```

Comment:
I could be misinterpreting what you mean here, but I don't think this is an attack vector. Or perhaps my comment is confusing. The Get function first looks in the catalog for an external plugin with the name provided, so one registered with that name by an administrator, then looks for builtin plugins. File names are a different parameter which are not used to search for plugins. So if an administrator wanted to say, patch a buggy internal plugin all they have to do is register a new external plugin with the same `plugin_name` as the builtin one. The backends would automatically start using the new plugin on next load without having to reconfigure all the backend plugins (though you might need to call reset on the plugins in the backend to actually stop the running plugin). Let me know if you think this is ok functionality.